cross-app-access-blog (#1603)
oktadeveloper/okta-blog: _source/_posts/2025-09-02-cross-app-access.md

---
layout: blog_post
title: "Build Secure Agent-to-App Connections with Cross App Access (XAA)"
author: sohail-pathan
by: advocate
communities: [javascript]
description: "Learn how to securely connect agents to applications using Cross App Access (XAA) in this comprehensive guide."
tags: [xaa, cross-app-access, enterprise-ai, agentic-ai, mcp, oauth, sso]
tweets:
- ""
- ""
- ""
- ""
image: blog/cross-app-access/social.jpg
type: conversion
---

Secure access with enterprise IT oversight between independent applications that communicate with each other is a recognized gap in [OAuth 2.0](https://developer.okta.com/docs/concepts/oauth-openid/). Enterprises can't effectively regulate cross-app communication, as OAuth 2.0 consent screens rely on users granting access to their individual accounts. Now, with the advent of AI agents that communicate across systems, the need to solve the gap is even greater.

> Cross App Access (XAA) is a new protocol that lets integrators enable secure agent-to-app and app-to-app access. Instead of scattered integrations and repeated logins, enterprise IT admins gain centralized control: they can decide what connects, enforce security policies, and see exactly what's being accessed. This unlocks seamless, scalable integrations across apps — whether it's just two like Google Calendar and Zoom, or hundreds across the enterprise. Read more about Cross App access in this post:

Reviewer: ```suggestion
Cross App Access (XAA) is a new protocol that lets integrators enable secure agent-to-app and app-to-app access. Instead of scattered integrations and repeated logins, enterprise IT admins gain centralized control: they can decide what connects, enforce security policies, and see exactly what's being accessed. This unlocks seamless, scalable integrations across apps — whether it's just two like Google Calendar and Zoom, or hundreds across the enterprise. Read more about Cross App Access in this post:
```

{% excerpt /blog/2025/06/23/enterprise-ai %}

> {% youtube 3VLzeT1EGrg %}

Reviewer: ```suggestion
Or watch the video about Cross App Access:

{% youtube 3VLzeT1EGrg %}
```

> In this post, we'll go hands-on with Cross App Access. Using **Todo0** (Resource App) and **Agent0** (Requesting App) as our sample applications, and **Okta as the enterprise Identity Provider (IdP)**, we'll show you how to set up trust, exchange tokens, and enable secure API calls between apps. By the end, you'll not only understand how the protocol works but also have a working example you can adapt to your own integrations.

Reviewer: ```suggestion
In this post, we'll go hands-on with Cross App Access. Using **Todo0** (Resource App) and **Agent0** (Requesting App) as our sample applications, and **Okta as the enterprise Identity Provider (IdP)**, we'll show you how to set up trust, exchange tokens, and enable secure API calls between apps that enables enterprise IT oversight. By the end, you'll not only understand how the protocol works but also have a working example you can adapt to your own integrations.
```

If you'd rather watch a video of the setup and how XAA works, check this one out.

{% youtube vi5JpbGRATE %}

> ## Prerequisites for Cross App Access (XAA) setup

Reviewer: ```suggestion
## Prerequisites to set up the AI agent to app using Cross App Access (XAA)
```

Adding in keywords like AI agent

To set up secure agent-to-app connections with Cross App Access (XAA), you'll need the following:

1. **Okta Developer Account (Integrator Free Plan):** You'll need an Okta Developer Account with the Integrator Free Plan. This account will act as your Identity Provider (IdP) for setting up Cross App Access.  
   * If you don't already have an account, sign up for a new one here: [Okta Integrator Free Plan](https://developer.okta.com/signup)  

>    * Once created, log in to your new org.

Reviewer: ```suggestion
   * Once created, sign in to your new org
```

We sign in, not log in

2. **AWS Credentials:** You'll need an **AWS Access Key ID** and **AWS Secret Access Key**.  
   * The IAM user or role associated with these credentials must have access to **Amazon Bedrock,** specifically the **Claude 3.7 Sonnet model,** enabled.  
   * If you don't know how to obtain the credentials, [follow this guide](https://github.com/oktadev/okta-cross-app-access-mcp/blob/main/guide/aws-bedrock.md).

3. **Developer Tools:** These tools are essential for cloning, editing, building, and running your demo applications:  
   * **[Git](https://git-scm.com/downloads)** – to clone and manage the repository.  
   * **[VS Code](https://code.visualstudio.com/Download)** – for reading and modifying the sample source code.  
   * **[Dev Containers Extension (VS Code)](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers)** – recommended, as it automatically configures dependencies and environments when you open the project.  
   * **[Docker](https://www.docker.com/products/docker-desktop/)** – required by the Dev Container to build and run the sample applications in isolated environments.

**Table of Contents**{: .hide }

* Table of Contents
{:toc}

> ## Create applications in the Okta Admin Console

Reviewer: ```suggestion
## Use Okta to secure AI applications with OAuth 2.0 and OpenID Connect (OIDC)
```

SEO changes

Before we dive into the code, we need to register our apps with Okta. In this demo:

> * **Agent0** → the AI agent **requesting app** (makes the API call on behalf of the user).  

Reviewer: ```suggestion
* **Agent0**: the AI agent **requesting app** (makes the API call on behalf of the user)
```

Bullet points don't end in periods unless it's multi-sentence per styles

* **Todo0** → the **resource app** (owns the protected API).  
* **Managed connection** → the trust relationship between the two apps, created in Okta.

We'll create both apps in your Okta Integrator Free Plan account, grab their client credentials, and then connect them.

> ### Enable Cross App Access (early access)

Reviewer: ```suggestion
### Enable Cross App Access in your Okta org
```

> ⚠️ **Note:** Cross App Access is currently a **self-service Early Access (EA) feature**. It must be enabled through the Admin Console before the apps appear in the catalog. If you don't see the option right away, refresh and confirm you have the necessary admin permissions. Learn more in the [Okta documentation on managing EA and beta features](https://help.okta.com/oie/en-us/content/topics/security/manage-ea-and-beta-features.htm).

> 1. Sign in to your Okta Developer Account.  

Reviewer: ```suggestion
1. Sign in to your Okta Integrator Free plan account 
```

2. In the **Okta Admin Console**, select **Settings → Features**.  
3. Navigate to **Early access**.  
4. Find **Cross App Access** and select **Turn on** (enable the toggle).  
5. Refresh the Admin Console.

> {% img blog/cross-app-access/image9.jpg alt:"Enable Cross App Access feature in Okta Admin Console" width:"782" %}{: .center-image }

Reviewer: Why not width of 800? this is our normal

Author: done.
